File: September-LeetCoding-Challenge/08-Sum-of-Root-To-Leaf-Binary-Numbers/Sum-of-Root-To-Leaf-Binary-Numbers.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public int SumRootToLeaf(TreeNode root) {
        if(root == null) return 0;
        
        return Traverse(root, 0);
    }
    
    private int Traverse(TreeNode node, int v){
        if(node == null) {
            return 0;
        }
        
        v = v * 2 + node.val;
        
        return node.left == node.right ? v : Traverse(node.left, v) + Traverse(node.right, v);
    }
}

File: September-LeetCoding-Challenge/08-Sum-of-Root-To-Leaf-Binary-Numbers/Sum-of-Root-To-Leaf-Binary-Numbers.rb
def sum_root_to_leaf(root)
    rec(root, 0)
end
    
def rec(root, val)
    if root.nil?
        return 0
    else
        val = val * 2 + root.val
        if root.left.nil? && root.right.nil?
           return val
        end
        rec(root.left, val) + rec(root.right, val)
        
    end
end